---
title: "Title"
author: "Author"
---

Copied from User Guide dated 22 April 2023.

# Introduction

# Preliminaries

# Quick Start

# Reading Microarray Data

# Quality Assessment

# Pre-Processing Two-Color Data

# Two-color Case Studies

## Swirl Zebrafish: A Single-Group Experiment

In this section we consider a case study in which two RNA sources are compared directly on a set of replicate or dye-swap arrays. The case study includes reading in the data, data display and exploration, as well as normalization and differential expression analysis. The analysis of differential expression is analogous to a classical one-sample test of location for each gene.

In this example we assume that the data is provided as a GAL file called `fish.gal` and raw SPOT output files and that these files are in the current working directory. The data used for this case study can be downloaded from `https://bioinf.wehi.edu.au/limmaGUI/DataSets.html`.

```{r}
dir("data/swirl")
```


**Background.** The experiment was carried out using zebrafish as a model organism to study the early development in vertebrates. Swirl is a point mutant in the BMP2 gene that affects the dorsal/ventral body axis. The main goal of the Swirl experiment is to identify genes with altered expression in the Swirl mutant compared to wild-type zebrafish.

**The hybridizations.** Two sets of dye-swap experiments were performed making a total of four repli- cate hybridizations. Each of the arrays compares RNA from swirl fish with RNA from normal (“wild type”) fish. The experimenters have prepared a tab-delimited targets file called `SwirlSamples.txt` which describes the four hybridizations:

```{r}
library(limma)
targets <- readTargets("data/swirl/SwirlSample.txt")
targets
```

We see that slide numbers 81, 82, 93 and 94 were used to make the arrays. On slides 81 and 93, swirl RNA was labeled with green (Cy3) dye and wild type RNA was labeled with red (Cy5) dye. On slides 82 and 94, the labelling was the other way around.

Each of the four hybridized arrays was scanned on an Axon scanner to produce a TIFF image, which was then processed using the image analysis software SPOT. The data from the arrays are stored in the four output files listed under `FileName`. Now we read the intensity data into an `RGList` object in R. The default for SPOT output is that `Rmean` and `Gmean` are used as foreground intensities and `morphR` and `morphG` are used as background intensities:

```{r}
RG <- read.maimages(targets, source = "spot", path = "data/swirl")
RG
```

**The arrays.** The microarrays used in this experiment were printed with 8448 probes (spots), including 768 control spots. The array printer uses a print head with a 4x4 arrangement of print-tips and so the microarrays are partitioned into a 4x4 grid of tip groups. Each grid consists of 22x24 spots that were printed with a single print-tip.

Unlike most image analysis software, SPOT does not store probe annotation in the output files, so we have to read it separately. The gene name associated with each spot is recorded in a GenePix array list (GAL) file:

```{r}
RG$genes <- readGAL("fish.gal", path = "data/swirl")
RG$genes[1:30, ]
```

Because we are using SPOT output, the 4x4x22x24 print layout also needs to be set. The easiest way to do this is to infer it from the GAL file:

```{r}
RG$printer <- getLayout(RG$genes)
```

Image plots. It is interesting to look at the variation of background values over the array. Consider image plots of the red and green background for the first array:

```{r}
imageplot(log2(RG$Rb[, 1]), RG$printer, low = "white", high = "red")
imageplot(log2(RG$Gb[, 1]), RG$printer, low = "white", high = "green")
```

Image plot of the un-normalized log-ratios or M-values for the first array:

```{r}
MA <- normalizeWithinArrays(RG, method = "none")
imageplot(MA$M[, 1], RG$printer, zlim = c(-3, 3))
```

The `imageplot` function lies the slide on its side, so the first print-tip group is bottom left in this plot. We can see a red streak across the middle two grids of the 3rd row caused by a scratch or dust on the array. Spots which are affected by this artifact will have suspect M-values. The streak also shows up as darker regions in the background plots.

**MA-plots.** An MA-plot plots the log-ratio of R vs G against the overall intensity of each spot. The log-ratio is represented by the M-value, M = log2(R) − log2(G), and the overall intensity by the A-value, A = (log2(R) + log2(G))/2. Here is the MA-plot of the un-normalized values for the first array:

```{r}
plotMD(MA)
```

The red streak seen on the image plot can be seen as a line of spots in the upper right of this plot. Now we plot the individual MA-plots for each of the print-tip groups on this array, together with the loess curves which will be used for normalization:

```{r}
plotPrintTipLoess(MA)
```

**Normalization.** Print-tip loess normalization:

```{r}
MA <- normalizeWithinArrays(RG)
plotPrintTipLoess(MA)
```

We have normalized the M-values with each array. A further question is whether normalization is required between the arrays. The following plot shows overall boxplots of the M-values for the four arrays.

```{r}
boxplot(MA$M ~ col(MA$M), names = colnames(MA$M))
```

There is evidence that the different arrays have different spreads of M-values, so we will scale normalize between the arrays.

```{r}
MA <- normalizeBetweenArrays(MA, method = "scale")
boxplot(MA$M ~ col(MA$M), names = colnames(MA$M))
```

Note that scale-normalization is not done routinely for all two-color data sets, in fact it is rarely done with newer platforms. However it does give good results on this data set. It should only be done when there is good evidence of a scale difference in the M-values.

**Linear model.** First setup an appropriate design matrix. The negative numbers in the design matrix indicate the dye-swaps:

```{r}
design <- modelMatrix(targets, ref = "wild type")
design
```

Now fit a simple linear model for each gene. This has the effect of estimating the average M-value for each gene, adjusting for the dye-swaps.

```{r}
fit <- lmFit(MA, design)
fit
```

In the above fit object, `coefficients` is the average M-value for each gene and `sigma` is the sample standard deviations for each gene. Ordinary *t*-statistics for comparing mutant to wt could be computed by

```{r, eval=FALSE}
ordinary.t <- fit$coefficients / fit$stdev.unscaled / fit$sigma
```

We prefer though to use empirical Bayes moderated *t*-statistics which are computed below.

Now create an mean difference plot displaying the log-fold-changes and average A-values for each
gene.

```{r}
plotMD(fit)
abline(0, 0, col = "blue")
```

**Empirical Bayes analysis.** We will now go on and compute empirical Bayes statistics for differential expression. The moderated *t*-statistics use sample standard deviations which have been squeezed towards a pooled standard deviation value.

```{r}
fit <- eBayes(fit)
qqt(fit$t, df = fit$df.prior + fit$df.residual, pch = 16, cex = 0.2)
abline(0, 1)
```

Visually there seems to be plenty of genes which are differentially expressed. We will obtain a summary table of some key statistics for the top genes.

```{r}
options(digits = 3)
topTable(fit, number = 30)
```

The top gene is BMP2 which is significantly down-regulated in the Swirl zebrafish, as it should be because the Swirl fish are mutant in this gene. Other positive controls also appear in the top 30 genes in terms.

In the table, `t` is the empirical Bayes moderated *t*-statistic, the corresponding P-values have been adjusted to control the false discovery rate and `B` is the empirical Bayes log odds of differential expression.

```{r}
plotMD(fit)
top30 <- order(fit$lods, decreasing = TRUE)[1:30]
text(fit$Amean[top30], fit$coefficients[top30], labels = fit$genes[top30, "Name"], cex = 0.8, col = "blue")
```

## Apoa1 Knockout Mice: A Two-Group Common-Reference Experiment

In this section we consider a case study where two RNA sources are compared through a common reference RNA. The analysis of the log-ratios involves a two-sample comparison of means for each gene.

In this example we assume that the data is available as an `RGList` in the data file `Apoa1.RData`. The data used for this case study can be downloaded from `https://bioinf.wehi.edu.au/limma`.

**Background.** The data is from a study of lipid metabolism by [3]. The apolipoprotein AI (Apoa1) gene is known to play a pivotal role in high density lipoprotein (HDL) metabolism. Mice which have the Apoa1 gene knocked out have very low HDL cholesterol levels. The purpose of this experiment is to determine how Apoa1 deficiency affects the action of other genes in the liver, with the idea that this will help determine the molecular pathways through which Apoa1 operates.

**Hybridizations.** The experiment compared 8 Apoa1 knockout mice with 8 normal C57BL/6 (”black six”) mice, the control mice. For each of these 16 mice, target mRNA was obtained from liver tissue and labelled using a Cy5 dye. The RNA from each mouse was hybridized to a separate microarray. Common reference RNA was labelled with Cy3 dye and used for all the arrays. The reference RNA was obtained by pooling RNA extracted from the 8 control mice.

```{r, echo=FALSE}
knitr::kable(
  data.frame(
    "Number of arrays" = rep(8, 2),
    "Red" = c('Normal "black six" mice', "Apoa1 knockout"),
    "Green" = rep("Pooled reference", 2),
    check.names = FALSE
  )
)
```

This is an example of a single comparison experiment using a common reference. The fact that the comparison is made by way of a common reference rather than directly as for the swirl experiment makes this, for each gene, a two-sample rather than a single-sample setup.

```{r}
load("data/Apoa1.RData")
# objects()
names(RG)
RG$targets
MA <- normalizeWithinArrays(RG)
cols <- MA$targets$Cy5
cols[cols == "C57BL/6"] <- "blue"
cols[cols == "ApoAI-/-"] <- "yellow"
boxplot(MA$M ~ col(MA$M), names = rownames(MA$targets), col = cols, xlab = "Mouse", ylab = "M-values")
```

Since the common reference here is a pool of the control mice, we expect to see more differences from the pool for the knock-out mice than for the control mice. In terms of the above plot, this should translate into a wider range of M-values for the knock-out mice arrays than for the control arrays, and we do see this. Since the different arrays are not expected to have the same range of M-values, between-array scale normalization of the M-values is not appropriate here.

Now we can go on to estimate the fold change between the two groups. In this case the design matrix has two columns. The coefficient for the second column estimates the parameter of interest, the log-ratio between knockout and control mice.

```{r}
design <- cbind("Control-Ref" = 1, "KO-Control" = MA$targets$Cy5 == "ApoAI-/-")
design
fit <- lmFit(MA, design)
fit$coef[1:5, ]
fit <- eBayes(fit)
options(digits = 3)
```

Normally at this point one would just type

```r
topTable(fit, coef = 2)
```

However, the gene annotation is a bit wide for the printed page, so we will tell `topTable()` to show just one column of the annotation information:

```{r}
topTable(fit, coef = 2, number = 15, genelist = fit$genes$NAME)
```

Notice that the top gene is Apoa1 itself which is heavily down-regulated. Theoretically the M-value should be minus infinity for Apoa1 because it is the knockout gene. Several of the other genes are closely related. The top eight genes here were confirmed by independent assay subsequent to the microarray experiment to be differentially expressed in the knockout versus the control line.

```{r}
volcanoplot(fit, coef = 2, highlight = 8, names = fit$genes$NAME, main = "KO vs Control")
```

## Weaver Mutant Mice: A Composite 2x2 Factorial Experiment

### Background

This case study considers a more involved two-color analysis in which the RNA sources have a factorial structure with two factors.

The study examined the development of neurons in wild-type and weaver mutant mice [7]. The weaver mutation affects cerebellar granule neurons, the most numerous cell-type in the central ner- vous system. Weaver mutant mice are characterized by a weaving gait. Granule cells are generated in the first postnatal week in the external granule layer of the cerebellum. In normal mice, the terminally differentiated granule cells migrate to the internal granule layer but in mutant mice the cells die before doing so, meaning that the mutant mice have strongly reduced numbers of cells in the internal granule layer. The expression level of any gene which is specific to mature granule cells, or is expressed in response to granule cell derived signals, is greatly reduced in the mutant mice.

### Sample Preparation and Hybridizations

At each time point (P11 = 11 days postnatal and P21 = 21 days postnatal) cerebella were isolated from two wild-type and two mutant littermates and pooled for RNA isolation. RNA was then divided into aliquots and labelled before hybridizing to the arrays. (This means that aliquots are technical replicates, arising from the same mice and RNA extraction. In our analysis here, we will ignore this complication and will instead treat the aliquots as if they were biological replicates. See Yang and Speed (2002) for a detailed discussion of this issue in the context of this experiment.) A pool of RNA was also made by combining the different RNA samples.

There are four different treatment combinations, P11wt, P11mt, P21wt and P21mt, comprising a 2x2 factorial structure. The RNA samples were hybridized to ten two-color microarrays, spotted with a 20k Riken clone library. There are six arrays comparing the four different RNA sources to the RNA pool, and four arrays making direct comparisons between the four treatment combinations.

The microarray images were scanned using SPOT image analysis software.

### Data input

The data used for this case study can be downloaded from `http://bioinf.wehi.edu.au/limma/ data/weaverfull.rar`. The data are provided courtesy of Drs Jean Yang and Elva Diaz.

First read in the targets frame:

```{r}
library(limma)
targets <- readTargets("data/weaverfull/targets.txt")
rownames(targets) <- removeExt(targets$FileName)
targets
```

Exploratory analysis showed that the segmented area for spots for these arrays was quite variable, with a median spot area just over 50 pixels. A small proportion of spots had very small segmented sizes, suggesting that the intensities for these spots might be unreliable. It was therefore decided to set a spot quality weight function, so any spot with an area less than 50 pixels will get reduced weight. The function is set so that any spot with zero area will get zero weight:

```{r}
wtfun <- function(x) pmin(x$area / 50, 1)
```

Then read the SPOT files containing the intensity data using file names recorded in the targets file. The data files are stored in the subdirectory `/spot`:

```{r}
RG <- read.maimages(targets, source = "spot", path = "data/weaverfull/spot", wt.fun = wtfun)
```

Finally, we set the print-tip layout. These arrays were printed using a print-head with 8 rows and 4 columns of print tips:

```{r}
RG$printer <- list(ngrid.r = 8, ngrid.c = 4, nspot.r = 25, nspot.c = 24)
```

### Annotation

Probe annotation is contained a separate file. The rows in the annotation file are as for the intensity data. Columns give Riken chip rearray IDs, GenBank accession numbers and UniGene information.

```{r}
Annotation <- read.delim("data/weaverfull/091701RikenUpdatev3.txt", 
                         comment.char = "", quote = "\"", check.names = FALSE, stringsAsFactors = FALSE)
names(Annotation)
```

For our purposes, we will keep the Riken IDs and GenBank accessions, putting these into the data object:

```{r}
RG$genes <- Annotation[, c(1, 2)]
colnames(RG$genes) <- c("RikenID", "GenBank")
```

Where possible, we find gene symbols corresponding to the GenBank accession numbers, by using the mouse organism package constructed from the NCBI database. Symbols can be found for only a little over 5000 of the probes.

```{r}
library(org.Mm.eg.db)
```

First we find the Entrez Gene ID for each accession number:

```{r}
EG.AN <- toTable(org.Mm.egACCNUM)
i <- match(RG$genes$GenBank, EG.AN[, "accession"])
EntrezID <- EG.AN[i, "gene_id"]
```

Then convert Entrez Gene IDs to symbols:

```{r}
EG.Sym <- toTable(org.Mm.egSYMBOL)
i <- match(EntrezID, EG.Sym[, "gene_id"])
RG$genes$Symbol <- EG.Sym[i, "symbol"]
```

### Quality Assessment and Normalization

We also read in a spot-types file and set a range of control spots.

```{r}
spottypes <- readSpotTypes("spottypes.txt", path = "data/weaverfull")
spottypes
RG$genes$Status <- controlStatus(spottypes, RG)
```

MA-plots were examined for all the arrays. Here we give the plot for array 9 only:

```{r}
# CSD note, points not coloured exactly as in user guide
plotMD(RG, column = 9, xlim = c(4, 15.5))
```


Here Buffer is an obvious negative control while 18S, GAPDH, Lysine, Threonine and Tubulin are single-gene positive controls, sometime called house-keeping genes. RikenTitration is a titration series of a pool of the entire Riken library, and can be reasonably expected to be non-differentially expressed. CerEstTitration is a titration of a pool of a cerebellum EST library. This will show higher expression in later mutant tissues. The Lys, Phe and Thr series are single-gene titration series which were not spiked-in in this case and can therefore be treated as negative controls.

The negative control probe intensities are quite high, especially for the red channel and especially for array 7:

```{r}
negative <- RG$genes$Status %in% c("Buffer", "LysTitration", "PheTitration", "ThrTitration")
par(mfrow = c(1, 2))
boxplot(log2(RG$G[negative, ]), las = 2, main = "Green background", ylab = "log2-intensity", col = "green")
boxplot(log2(RG$R[negative, ]), las = 2, main = "Red background", ylab = "log2-intensity", col = "red")
par(mfrow = c(1,1))
```

Later on, we will investigate setting array quality weights.

Now normalize the data. The Riken titration library, being based on a pool of a large number
of non-specific genes, should not be differentially expressed. We can take advantage of this by upweighting these probes in the print-tip normalization step. Here we give double-weight to the titration library probes, although higher weights could also be considered:

```{r}
w <- modifyWeights(RG$weights, RG$genes$Status, "RikenTitration", 2)
MA <- normalizeWithinArrays(RG, weights = w)
```

### Setting Up the Linear Model

The experiment has a composite design, with some arrays comparing back to the RNA pool as a common reference, and other arrays making direct comparisons between the treatment conditions of interest. The simplest design matrix is that which compares all the RNA samples back to the RNA pool.

```{r}
design <- modelMatrix(targets, ref = "Pool")
```

We also add an intercept term to extract probe-specific dye effects:

```{r}
design <- cbind(Dye = 1, design)
design
```

### Probe Filtering and Array Quality Weights

First we remove control probes, leaving only the regular probes of the Riken library:

```{r}
regular <- MA$genes$Status == "Riken"
MA2 <- MA[regular, ]
MA2$genes$Status <- NULL
```

Then we estimate array quality weights:

```{r}
aw <- arrayWeights(MA2, design)
options(digits = 3)
aw
```

The array weights multiply the spot weights already in the data object:

```{r}
library(statmod)
w <- matvec(MA2$weights, aw)
```

### Differential expression

Fit the linear model:

```{r}
fit <- lmFit(MA2, design, weights = w)
```

Now extract all possible comparisons of interest as contrasts. We look for the mutant vs wt comparisons at 11 and 21 days, the time effects for mutant and wt, and the interaction terms:

```{r}
cont.matrix <- makeContrasts(
  WT11.MT11 = P11mt - P11wt,
  WT21.MT21 = P21mt - P21wt,
  WT11.WT21 = P21wt - P11wt,
  MT11.MT21 = P21mt - P11mt,
  Int = (P21mt - P11mt) - (P21wt - P11wt),
  levels = design
)
fit2 <- contrasts.fit(fit, cont.matrix)
fit2 <- eBayes(fit2)
```

Adjustment for multiple testing, using Benjamini and Hochberg’s method to control the false dis- covery rate at 5% across all genes and all contrasts, leads to the following:

```{r}
results <- decideTests(fit2, method = "global")
summary(results)
```

The probes that show significant interactions are those which develop differently in the mutant compared to the wild-type between days 11 and 21. To see these:

```{r}
topTable(fit2, coef = "Int")
sessionInfo()
```

## Bob1 Mutant Mice: Arrays with Duplicate Spots

In this section we consider a case study in which all genes (ESTs and controls) are printed more than once on the array. This means that there is both within-array and between-array replication for each gene. The structure of the experiment is therefore essentially a randomized block experiment for each gene. The approach taken here is to estimate a common correlation for all the genes for between within-array duplicates. The theory behind the approach is explained in [32]. This approach assumes that all genes are replicated the same number of times on the array and that the spacing between the replicates is entirely regular.

In this example we assume that the data is available as an `RGList`.

**Background.** This data is from a study of transcription factors critical to B cell maturation by Lynn Corcoran and Wendy Dietrich at the WEHI. Mice which have a targeted mutation in the Bob1 (aka Pou2af1 or OBF-1) transcription factor display a number of abnormalities in the B lymphocyte compartment of the immune system. Immature B cells that have emigrated from the bone marrow fail to differentiate into full fledged B cells, resulting in a notable deficit of mature B cells.

**Arrays.** Arrays were printed at the Australian Genome Research Facility with expressed sequence tags (ESTs) from the National Institute of Aging 15k mouse clone library, plus a range of positive, negative and calibration controls. The arrays were printed using a 48 tip print head and 26x26 spots in each tip group. Data from 24 of the tip groups are given here. Every gene (ESTs and controls) was printed twice on each array, side by side by rows. The NIA15k probe IDs have been anonymized in the output presented here.

**Hybridizations.** A retrovirus was used to add Bob back to a Bob deficient cell line. Two RNA sources were compared using 2 dye-swap pairs of microarrays. One RNA source was obtained from the Bob deficient cell line after the retrovirus was used to add GFP (”green fluorescent protein”, a neutral protein). The other RNA source was obtained after adding both GFP and Bob protein. RNA from Bob+GFP was labelled with Cy5 in arrays 2 and 4, and with Cy3 in arrays 1 and 4.

**Image analysis.** The arrays were image analyzed using SPOT with “morph” background estima- tion.

The data used for this case study can be downloaded from `http://bioinf.wehi.edu.au/limma/`. The file should be placed in the working directory of your R session. (This case study was last updated on 29 June 2006 using R 2.3.0 and limma 2.7.5.)

```{r}
library(limma)
load("data/Bob.RData")
# objects()
design
names(RG)
RG$genes[1:40, ]
```

Although there are only four arrays, we have a total of eight spots for each gene, and more for the controls. Naturally the two M-values obtained from duplicate spots on the same array are highly correlated. The problem is how to make use of the duplicate spots in the best way. The approach taken here is to estimate the spatial correlation between the adjacent spots using REML and then to conduct the usual analysis of the arrays using generalized least squares.

First normalize the data using print-tip loess regression. The SPOT morph background ensures that the default background subtraction can be used without inducing negative intensities.

```{r}
MA <- normalizeWithinArrays(RG)
```

Then remove the control probes:

```{r}
MA2 <- MA[MA$genes$Library == "NIA15k", ]
```

Now estimate the spatial correlation. We estimate a correlation term by REML for each gene, and then take a trimmed mean on the atanh scale to estimate the overall correlation. This command will probably take at least a few minutes depending on the speed of your computer.

```{r}
options(digits = 3)
corfit <- duplicateCorrelation(MA2, design, ndups = 2) # A slow computation!
corfit$consensus.correlation
boxplot(tanh(corfit$atanh.correlations))
fit <- lmFit(MA2, design, ndups = 2, correlation = corfit$consensus)
fit <- eBayes(fit)
topTable(fit, n = 30)
volcanoplot(fit)
```
